tiledb/sm/array_schema/test/unit_attribute.cc

```diff
@@ -0,0 +1,207 @@
+/**
+ * @file tiledb/sm/array_schema/test/unit_arribute.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file contains unit tests for the attribute class
+ */
+
+#include <tdb_catch.h>
+
+#include "tiledb/sm/array_schema/attribute.h"
+
+using namespace tiledb::common;
+using namespace tiledb::sm;
+
+// Datatype::ANY attribute is var-sized by default
+TEST_CASE("ANY Attribute default cell val num", "[array_schema][attribute]") {
+  Attribute a("a", Datatype::ANY, false);
+  CHECK(a.cell_val_num() == constants::var_num);
+}
+
+// Pass non-default cell val num to constructor
+TEST_CASE(
+    "ANY Attribute construct non-default cell val num",
+    "[array_schema][attribute]") {
+  CHECK_THROWS(Attribute("a", Datatype::ANY, 1, DataOrder::UNORDERED_DATA));
+  CHECK_THROWS(Attribute("a", Datatype::ANY, 2, DataOrder::UNORDERED_DATA));
+
+  REQUIRE_NOTHROW(Attribute(
+      "a", Datatype::ANY, constants::var_num, DataOrder::UNORDERED_DATA));
+  Attribute a(
+      "a", Datatype::ANY, constants::var_num, DataOrder::UNORDERED_DATA);
+  CHECK(a.cell_val_num() == constants::var_num);
+}
+
+// Set cell val num after construction
+TEST_CASE("ANY Attribute set_cell_val_num", "[array_schema][attribute]") {
+  Attribute a("a", Datatype::ANY, false);
+  CHECK(a.cell_val_num() == constants::var_num);
+
+  CHECK_THROWS(a.set_cell_val_num(1));
+  CHECK_THROWS(a.set_cell_val_num(2));
+  CHECK_NOTHROW(a.set_cell_val_num(constants::var_num));
+  CHECK(a.cell_val_num() == constants::var_num);
```

Comment:
This all looks like it needs `SECTION` and possibly `GENERATE`. For this function specifically, I'm not too worried about it because is slated for removal as part of the C.41 effort.